hyperspy/learn/mva.py

```diff
@@ -1636,7 +1642,747 @@ def undo_treatments(self):
+        idx = np.argsort(clustersizes)[::-1]
+        lut = np.zeros_like(idx)
+        lut[idx] = np.arange(n_clusters)
+        sorted_labels = lut[labels]
+        shape = (n_clusters, self.data.shape[0])
+        cluster_labels = np.full(shape, np.nan)
+        sorted_cluster_centers = np.zeros_like(cluster_centers)            
+        # now create the labels from these sorted labels
+        for i in range(n_clusters):
+            cluster_labels[i, navigation_mask] = \
+                np.where(sorted_labels == i, 1, 0)
+            sorted_cluster_centers[i,:] = cluster_centers[lut[i],:]\
+                /clustersizes[lut[i]]                
+
+
+
+        return sorted_labels,cluster_labels, sorted_cluster_centers
+
+
+    def cluster_analysis(self,
+                         n_clusters=None,
+                         scaling="norm",
+                         scaling_kwargs={},
+                         use_decomposition_results=True,
+                         use_decomposition_results_for_centers=False,
+                         number_of_components=None,
+                         navigation_mask=None,
+                         signal_mask=None,
+                         algorithm='kmeans',
+                         return_info=False,
+                         **kwargs):
+        """
+        Cluster analysis of a signal or decomposition results of a signal
+
+        Results are stored in `learning_results`.
+
+        Parameters
+        ----------
+        n_clusters : int
+            Number of clusters to find.
+        scaling : {"standard","norm","minmax",None or scikit learn scaling method}
+            default: 'norm'
+            Preprocessing the data before cluster analysis requires scaling
+            the data to be clustered to similar scales. Standard scaling
+            adjusts each feature to have uniform variation. Norm scaling
+            adjusts treats the set of features like a vector and
+            each measurement is scaled to length 1.
+            You can also pass one of the scikit-learn preprocessing
+            scale_method = import sklearn.processing.StandadScaler
+            scaling = scale_method
+            See scaling methods in scikit-learn preprocessing for further
+            details.
+        scaling_kwargs : dict
+            Additional parameters passed to the cluster scaling algorithm.
+            See sklearn.preprocessing scaling methods for further details
+        use_decomposition_results : {True,False,"decomposition","bss"}
+            If True or "decomposition" the signal's decomposition results are used
+            for clustering. If "bss" the blind source separation results are used
+            If False the raw data is used Note that this can be memory intensive 
+            and is only recommened if the Signal has a small signal
+            or navigation dimensions.
+        use_decomposition_results_for_centers : {True,False,"decomposition","bss"}
```

Comment:
Similarly, this one could be renamed to something like `computer_center_signals` and take `{False, "decomposition", "signal", BaseSignal}` with `BaseSignal` being any signal with the right dimensions.

Reply:
but to be clear...
use_decomposition_results, becomes X  - user has to enter something
use_decomposition_results_for_centers - leave as a keyword or also set as an argument user has to enter something ?

Reply:
What do you think about:

```
if use_decomposition_results is None:
    use_decomposition_results = X if X == "decomposition" else "signal"
```

By the way, if you take into account my suggestion, ``use_decomposition_results`` may need a new name.

Reply:
I've made quite a few changes:
Custom clustering algorithm or scaling method can be passed as per decomposition method Custom scaling was possible but this is now standard and consistent with the decomposition method
use_decomposition_ results  is replaced with cluster_source which has to be "bss", "decomposition", "signal" - user has to supply an entry
use_decomposition_results_for_centers becomes cluster_source_for_centers and can be "bss","decomposition", BaseSignal or None - if None it is the same as cluster_source
Tests have been updated but I've not updated the documentation until the code is reviewed.

Reply:
The checks for BaseSignal also seem clunky - might be neater if people could pass a suitable np or dask array instead..

Reply:
The tests currently fail on test_plot_cluster_centers_nav2_sig2 but seems to be variations in the dependencies - I can get some tests to pass and some to fail depending on how I setup the env or build on windows or linux. I will comment it out for now

Reply:
I've updated to allow for "signal","bss","decomposition" or BaseSignal as a cluster source and source for cluster_centers.  To neaten the code and generalise I created a _get_cluster_signal which handles returning the signal to process at different steps. This took some re-organization so the code is different but more general. 
 To allow for custom or different sklearn methods I followed the decomposition method which checks for "fit" or "fit_transform" attr and I also check for "labels_" in the attr after clustering. 
I've increased the number of tests to cover exceptions and the various options for Signals as cluster_source or source_for centers.
I've changed the documentation to match and included a brief example of how you might use these.